File: app/src/main/java/org/nitri/opentopo/CacheSettingsFragment.java
package org.nitri.opentopo;

import android.annotation.SuppressLint;
import android.app.AlarmManager;
import android.app.Dialog;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import org.osmdroid.config.Configuration;

import java.io.File;

public class CacheSettingsFragment extends DialogFragment {

    private static final String TAG = CacheSettingsFragment.class.getSimpleName();

    // Hardcoded in org.osmdroid.config.DefaultConfigurationProvider
    private static final String DEFAULT_TILE_CACHE = "osmdroid/tiles";
    private static final int DEFAULT_CACHE_SIZE = 600;

    private final static String PREF_TILE_CACHE = "tile_cache";
    private final static String PREF_CACHE_SIZE = "cache_size";

    private EditText etTileCache;
    private EditText etCacheSize;


    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(requireActivity());
        LayoutInflater inflater = requireActivity().getLayoutInflater();

        @SuppressLint("InflateParams") View view = inflater.inflate(R.layout.fragment_cache_settings, null);

        SharedPreferences prefs = requireActivity().getSharedPreferences(MapFragment.MAP_PREFS, Context.MODE_PRIVATE);

        TextView tvExternalStorageRoot = view.findViewById(R.id.tvExternalStorageRoot);

        etTileCache = view.findViewById(R.id.etTileCache);
        etCacheSize = view.findViewById(R.id.etCacheSize);

        String storageRoot = Configuration.getInstance().getOsmdroidBasePath().getAbsolutePath();
        tvExternalStorageRoot.setText(getString(R.string.storage_root, storageRoot));

        String currentTileCache = prefs.getString(PREF_TILE_CACHE, DEFAULT_TILE_CACHE);
        int currentCacheSize = prefs.getInt(PREF_CACHE_SIZE, DEFAULT_CACHE_SIZE);

        etTileCache.setText(currentTileCache);
        etCacheSize.setText(String.valueOf(currentCacheSize));

        builder.setView(view)
                .setPositiveButton(android.R.string.ok, (dialog, id) -> {
                    String newTileCache = etTileCache.getText().toString();
                    int newCacheSize = Integer.parseInt(etCacheSize.getText().toString());
                    SharedPreferences.Editor editor = prefs.edit();
                    editor.putString(PREF_TILE_CACHE, newTileCache);
                    editor.putInt(PREF_CACHE_SIZE, newCacheSize);
                    editor.apply();
                    File cacheDir = new File(storageRoot + "/" + newTileCache);
                    if (cacheDir.mkdirs()) {
                        Log.i(TAG, "Tile cache created: " + newTileCache);
                    }
                    Configuration.getInstance().setOsmdroidTileCache(cacheDir);
                    Configuration.getInstance().setTileFileSystemCacheMaxBytes((long) newCacheSize * 1024 * 1024);
                    Configuration.getInstance().save(requireActivity().getApplicationContext(), prefs);
                    if (!newTileCache.equals(currentTileCache) || newCacheSize != currentCacheSize)
                        restart();
                    dismiss();
                })
                .setNegativeButton(android.R.string.cancel, (dialog, id) -> dismiss())
                .setNeutralButton(R.string.reset, (dialog, id) -> {
                    // NOP: override in onResume() (do not dismiss dialog)
                });


        AlertDialog dialog = builder.create();
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);

        return dialog;
    }

    /**
     * Restart the rude way
     */
    private void restart() {
        requireActivity().finish();
        startActivity(requireActivity().getIntent());
        requireActivity().overridePendingTransition(0, 0);
    }

    @Override
    public void onResume() {
        super.onResume();
        final AlertDialog dialog = (AlertDialog) getDialog();
        if (dialog != null) {
            Button positiveButton = dialog.getButton(Dialog.BUTTON_NEUTRAL);
            positiveButton.setOnClickListener(v -> {
                etTileCache.setText(DEFAULT_TILE_CACHE);
                etCacheSize.setText(String.valueOf(DEFAULT_CACHE_SIZE));
            });
        }
    }

}

File: app/src/main/java/org/nitri/opentopo/CacheSettingsFragment.kt
package org.nitri.opentopo

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.view.Window
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.core.content.edit
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.setViewTreeLifecycleOwner
import androidx.lifecycle.setViewTreeViewModelStoreOwner
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.preference.PreferenceManager
import androidx.savedstate.setViewTreeSavedStateRegistryOwner
import org.nitri.opentopo.analytics.AnalyticsNames
import org.nitri.opentopo.analytics.AnalyticsProvider
import org.nitri.opentopo.ui.theme.OpenTopoTheme
import org.nitri.opentopo.util.Utils
import org.osmdroid.config.Configuration
import java.io.File


class CacheSettingsFragment : DialogFragment() {

    private var externalStorageState by mutableStateOf(false)
    private var tileCacheState by mutableStateOf("")
    private var cacheSizeState by mutableStateOf("")

    private var currentExternalStorage = false
    private var currentTileCache = ""
    private var currentCacheSize = 0

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val fragmentActivity = activity ?: return super.onCreateDialog(savedInstanceState)
        AnalyticsProvider.get(fragmentActivity).trackScreen(
            AnalyticsNames.Screen.CACHE_SETTINGS,
            CacheSettingsFragment::class.java.simpleName
        )

        val defaultPrefs = PreferenceManager.getDefaultSharedPreferences(fragmentActivity.applicationContext)
        val cachePrefs: SharedPreferences = fragmentActivity.getSharedPreferences("cache_prefs", Context.MODE_PRIVATE)

        currentExternalStorage = cachePrefs.getBoolean(PREF_EXTERNAL_STORAGE,
            defaultPrefs.getBoolean(PREF_EXTERNAL_STORAGE, false)
        )
        currentTileCache = defaultPrefs.getString(PREF_TILE_CACHE, DEFAULT_TILE_CACHE) ?: DEFAULT_TILE_CACHE
        currentCacheSize = defaultPrefs.getInt(PREF_CACHE_SIZE, DEFAULT_CACHE_SIZE)

        // Initialize state
        externalStorageState = currentExternalStorage
        tileCacheState = currentTileCache
        cacheSizeState = currentCacheSize.toString()

        val composeView = ComposeView(fragmentActivity).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                OpenTopoTheme(dynamicColor = false) {
                    val currentStorageRoot = Utils.getOsmdroidBasePath(fragmentActivity.applicationContext, externalStorageState).absolutePath
                    val storageRootText = stringResource(id = R.string.storage_root, currentStorageRoot)
                    CacheSettingsContent(
                        storageRootText = storageRootText,
                        externalStorage = externalStorageState,
                        onExternalStorageChange = { externalStorageState = it },
                        tileCache = tileCacheState,
                        onTileCacheChange = { tileCacheState = it },
                        cacheSize = cacheSizeState,
                        onCacheSizeChange = { cacheSizeState = it }
                    )
                }
            }
        }

        // Set the ViewTree owners on the ComposeView
        composeView.setViewTreeLifecycleOwner(this)
        composeView.setViewTreeViewModelStoreOwner(this)
        composeView.setViewTreeSavedStateRegistryOwner(this)

        val builder = AlertDialog.Builder(fragmentActivity)
            .setView(composeView)
            .setPositiveButton(android.R.string.ok, null) // Listener set in onResume to prevent auto-dismiss
            .setNegativeButton(android.R.string.cancel) { _: DialogInterface?, _: Int -> dismiss() }
            .setNeutralButton(R.string.reset, null) // Listener set in onResume

        val dialog = builder.create()
        setCancelable(false)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)

        dialog.setOnShowListener {
            val decorView = dialog.window?.decorView
            if (decorView != null) {
                // Compose can crash with 'ViewTreeLifecycleOwner not found from AlertDialogLayout'
                // if we don't set these on the dialog's decor view.
                decorView.setViewTreeLifecycleOwner(this)
                decorView.setViewTreeViewModelStoreOwner(this)
                decorView.setViewTreeSavedStateRegistryOwner(this)
            }

            // Required for Compose TextFields inside AppCompat AlertDialog.
            // Without this, Compose receives focus but the soft keyboard may not appear.
            dialog.window?.clearFlags(WindowManager.LayoutParams.FLAG_ALT_FOCUSABLE_IM)
            dialog.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_NOTHING)
        }

        return dialog
    }

    @Composable
    private fun CacheSettingsContent(
        storageRootText: String,
        externalStorage: Boolean,
        onExternalStorageChange: (Boolean) -> Unit,
        tileCache: String,
        onTileCacheChange: (String) -> Unit,
        cacheSize: String,
        onCacheSizeChange: (String) -> Unit
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Text(
                text = storageRootText,
                style = MaterialTheme.typography.bodyMedium
            )
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(
                    text = stringResource(id = R.string.external_storage),
                    modifier = Modifier.weight(1f),
                    style = MaterialTheme.typography.bodyLarge
                )
                Switch(
                    checked = externalStorage,
                    onCheckedChange = onExternalStorageChange
                )
            }
            OutlinedTextField(
                value = tileCache,
                onValueChange = onTileCacheChange,
                label = { Text(stringResource(id = R.string.tile_cache)) },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true
            )
            OutlinedTextField(
                value = cacheSize,
                onValueChange = onCacheSizeChange,
                label = { Text(stringResource(id = R.string.max_cache_size)) },
                modifier = Modifier.fillMaxWidth(),
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                singleLine = true
            )
        }
    }

    override fun onResume() {
        super.onResume()
        val dialog = dialog as? AlertDialog
        dialog?.let { ad ->
            ad.getButton(AlertDialog.BUTTON_NEUTRAL).setOnClickListener {
                tileCacheState = DEFAULT_TILE_CACHE
                cacheSizeState = DEFAULT_CACHE_SIZE.toString()
            }
            ad.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
                applySettings()
            }
        }
    }

    private fun applySettings() {
        val fragmentActivity = activity ?: return
        val appContext = fragmentActivity.applicationContext
        val defaultPrefs = PreferenceManager.getDefaultSharedPreferences(appContext)
        val cachePrefs: SharedPreferences = fragmentActivity.getSharedPreferences("cache_prefs", Context.MODE_PRIVATE)

        val newExternalStorage = externalStorageState
        val newTileCache = tileCacheState
        val newCacheSizeText = cacheSizeState
        val newCacheSize = newCacheSizeText.toIntOrNull() ?: -1

        if (newCacheSize == -1) {
            Log.e(TAG, "Invalid cache size: $newCacheSizeText")
        }

        if (newCacheSize > 0) {
            val newBasePath = Utils.getOsmdroidBasePath(appContext, newExternalStorage)
            val newCacheDir = File(newBasePath, newTileCache)

            // Validate the cache directory before applying.
            if (!Utils.isCacheDirValid(newCacheDir)) {
                Toast.makeText(appContext, R.string.storage_not_accessible, Toast.LENGTH_LONG).show()
                return
            }

            defaultPrefs.edit().apply {
                putString(PREF_TILE_CACHE, newTileCache)
                putInt(PREF_CACHE_SIZE, newCacheSize)
                apply()
            }
            cachePrefs.edit { putBoolean(PREF_EXTERNAL_STORAGE, newExternalStorage) }

            val configuration = Configuration.getInstance()
            configuration.osmdroidBasePath = newBasePath
            configuration.osmdroidTileCache = newCacheDir
            configuration.tileFileSystemCacheMaxBytes =
                newCacheSize.toLong() * 1024 * 1024
            // Remove any leftover sqlite files to avoid SqlTileWriter using them
            Utils.clearOsmdroidSqliteCache(appContext)
            configuration.save(appContext, defaultPrefs)
            val intent = Intent(ACTION_CACHE_CHANGED)
            val localBroadcastManager = LocalBroadcastManager.getInstance(
                fragmentActivity
            )
            if (currentExternalStorage != newExternalStorage || currentTileCache != newTileCache || currentCacheSize != newCacheSize) {
                localBroadcastManager.sendBroadcast(intent)
                fragmentActivity.finish()
            }
            dismiss()
        } else {
            Toast.makeText(context ?: fragmentActivity, R.string.invalid_cache_size, Toast.LENGTH_SHORT).show()
        }
    }

    companion object {
        private val TAG = CacheSettingsFragment::class.java.simpleName

        const val DEFAULT_TILE_CACHE =
            "tiles"  // Hardcoded in org.osmdroid.config.DefaultConfigurationProvider
        const val DEFAULT_CACHE_SIZE = 600
        const val PREF_TILE_CACHE = "tile_cache"
        const val PREF_CACHE_SIZE = "cache_size"
        const val PREF_EXTERNAL_STORAGE = "external_storage"
        const val ACTION_CACHE_CHANGED = "cache_changed"
    }
}
